medusa/tv/episode.py

```diff
@@ -87,20 +84,209 @@
+
+
+    @property
+    def season(self):
+        return NotImplemented
+
+    @season.setter
+    def season(self, value):
+        try:
+            assert value in (None, NotImplemented)
+        except AssertionError:
+            raise InvalidNumbering(
+                    'Invalid season for {0}: {1!r}'.format(
+                        self.__class__.__name__,
+                        value,
+                    )
+                )
+
+    @property
+    def episode(self):
+        return getattr(self, '_episode')
+
+    @episode.setter
+    def episode(self, value):
+        try:
+            setattr(self, '_episode', int(value))
+        except TypeError:
+            raise InvalidNumbering(
+                'Invalid season for {0}: {1!r}'.format(
+                    self.__class__.__name__,
+                    value,
+                )
+            )
+
+    def __str__(self):
+        """Magic method."""
+        return 'e{0:02d}'.format(self.episode)
+
+    @classmethod
+    def parse(cls, number, regex=None):
+        regex = regex or cls.regex
+        return regex.match(number).groupdict()
+
+
+class RelativeNumber(AbsoluteNumber):
+    """Standard episode number with season and episode."""
+
+    _attrs = 'season', 'episode'
+
+    def __init__(self, season, episode):
+        """
+        Initialize a relative numbering for an episode.
+
+        :param season:
+        :type season: int
+        :param episode:
+        :type episode: int
+        """
+        super(RelativeNumber, self).__init__(episode)
+
+    @property
+    def season(self):
+        return getattr(self, '_season')
+
+    @season.setter
+    def season(self, value):
+        try:
+            setattr(self, '_season', int(value))
+        except TypeError:
+            raise InvalidNumbering(
+                'Invalid season for {0}: {1!r}'.format(
+                    self.__class__.__name__,
+                    value,
+                )
+            )
+
+    def __str__(self):
+        """Magic method."""
+        return 's{0:02d}e{1:02d}'.format(self.season, self.episode)
+
+
+class AirByDateNumber(EpisodeNumber):
+    """Air-by-date episode number."""
+
+    _attrs = ['air_date']
+    date_fmt = '%Y-%m-%d'
+
+    def __init__(self, air_date):
+        """
+        Initialize an air-by-date numbering for an episode.
+
+        :param air_date:
+        :type air_date: datetime
+        """
+        self.air_date = air_date
+
+    def __str__(self):
+        """Magic method."""
+        return self.air_date.strftime(self.date_fmt)
+
+    @property
+    def air_date(self):
+        return getattr(self, '_air_date')
+
+    @air_date.setter
+    def air_date(self, value):
+        if not isinstance(value, (date, datetime)):
+            value = datetime.strptime(value, self.date_fmt)
+        with supress(AttributeError):
+            value = value.date()
+        setattr(self, '_air_date', value)
+
+    @classmethod
+    def parse(cls, number, fmt=None):
```

Comment:
I'm still undecided if I want to allow a `fmt` argument here (as you could always change the date_fmt in the class or instance).